### PR TITLE
[WIP] business_requirement_crm

### PR DESCRIPTION
**Pull request by martintorresvm**
This is the new PR which replaces the old one: https://github.com/OCA/project/pull/143

---

Depends on:
- [ ] business_requirement_deliverable #3
  <a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/OCA/business-requirement/pull/8%23discussion_r68704110%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/8%23discussion_r68704201%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/8%23discussion_r74006969%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/8%23discussion_r74026036%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/8%23discussion_r74038780%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/8%23issuecomment-246562807%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/8%23issuecomment-246643889%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/8%23issuecomment-247467060%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/8%23issuecomment-247474023%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/8%23issuecomment-247844044%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/8%23issuecomment-247915041%22%2C%20%22https%3A//github.com/OCA/business-requirement/pull/8%23discussion_r79345837%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/8%23issuecomment-246562807%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%5Cn%5B%21%5BCoverage%20Status%5D%28https%3A//coveralls.io/builds/7856217/badge%29%5D%28https%3A//coveralls.io/builds/7856217%29%5Cn%5CnCoverage%20remained%20the%20same%20at%2080.0%25%20when%20pulling%20%2A%2A9ca75b8ad6a4edd84e8bffd3623a564a8cd192f0%20on%20victormartinelicocorp%3A6_business_requirement_crm%2A%2A%20into%20%2A%2Aaf25af1b54bfc4fa2f871f713be1a340bc62aa69%20on%20OCA%3A8.0%2A%2A.%5Cn%22%2C%20%22created_at%22%3A%20%222016-09-13T03%3A25%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/2354108%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/coveralls%22%7D%7D%2C%20%7B%22body%22%3A%20%22%5Cn%5B%21%5BCoverage%20Status%5D%28https%3A//coveralls.io/builds/7860396/badge%29%5D%28https%3A//coveralls.io/builds/7860396%29%5Cn%5CnCoverage%20decreased%20%28-11.0%25%29%20to%2080.0%25%20when%20pulling%20%2A%2A62ee1a923a26fbc1e8e38bb47965d18208b91756%20on%20victormartinelicocorp%3A6_business_requirement_crm%2A%2A%20into%20%2A%2Aa59878447383fc5ef273ad0b54fb2f95cb7ba3f9%20on%20OCA%3A8.0%2A%2A.%5Cn%22%2C%20%22created_at%22%3A%20%222016-09-13T10%3A45%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/2354108%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/coveralls%22%7D%7D%2C%20%7B%22body%22%3A%20%22Needs%20a%20cleanup%20on%20the%20README%2C%20like%20in%20the%20other%20modules.%22%2C%20%22created_at%22%3A%20%222016-09-15T21%3A57%3A28Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1246629%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dreispt%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yes%3A%20PR%20is%20pending%20in%20the%20fork.%20We%20will%20get%20to%20it%20this%20week%20end%20%28almost%20there%21%29%22%2C%20%22created_at%22%3A%20%222016-09-15T22%3A34%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%2C%20%7B%22body%22%3A%20%22%40elicoidal%22%2C%20%22created_at%22%3A%20%222016-09-18T12%3A09%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/16461806%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/victormartinelicocorp%22%7D%7D%2C%20%7B%22body%22%3A%20%22%40elicoidal%20%40dreispt%20%22%2C%20%22created_at%22%3A%20%222016-09-19T05%3A49%3A28Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/16461806%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/victormartinelicocorp%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%2032c991a5fd1ff2ca7dd38044bc217312fdcfb2db%20business_requirement_crm/models/business.py%2022%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/8%23discussion_r74006969%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%40elicoidal%20This%20is%20also%20deprecated%20the%20%60br_ids%60%20feature%20is%20no%20longer%20introduced%20by%20base%20module%20%60business_requirement%60.%20I%20guess%20I%20should%20make%20this%20module%20depends%20on%20the%20%60business_requirement_project%60.%20What%20do%20you%20think%3F%22%2C%20%22created_at%22%3A%20%222016-08-09T07%3A03%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/16461806%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/victormartinelicocorp%22%7D%7D%2C%20%7B%22body%22%3A%20%22AFAIR%2C%20project%20dependency%20has%20been%20left%20in%20business_requirement%20module%20no%3F%5Cr%5CnI%20dont%20think%20this%20is%20a%20good%20idea%20to%20this%20particular%20module%20to%20depend%20on%20br_project%20%28which%20is%20specific%20for%20project%20management.%5Cr%5Cn%20I%20think%20this%20module%20should%20depend%20on%20br_deliverable%20so%20you%20might%20want%20to%20add%20some%20fields%20in%20br_deliverable%20if%20necessary%20%28or%20add%20them%20in%20this%20module%29%22%2C%20%22created_at%22%3A%20%222016-08-09T09%3A27%3A40Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%2C%20%7B%22body%22%3A%20%22Is%20also%20a%20little%20far%20for%20me%20in%20memory.%20But%20yes%20I%20remember%20that%20we%20decided%20left%20master_project%20on%20the%20business_requirement.%20%5Cr%5CnThen%20we%20might%20need%20to%20modify%20project%20to%20add%20one2many%20relation%20to%20the%20%60business.requirement%60%20so%20I%27ll%20do%20as%20you%20suggest%20and%20modify%20the%20br_deliverable.%5Cr%5CnAfter%20that%20I%27ll%20rebase%20this%20PR.%20%22%2C%20%22created_at%22%3A%20%222016-08-09T11%3A13%3A06Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/16461806%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/victormartinelicocorp%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20business_requirement_crm/models/business.py%3AL1-32%22%7D%2C%20%22Pull%20ae8704aac1972580699c69be6f8345e5237a9807%20business_requirement_deliverable_crm/models/crm.py%2027%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/8%23discussion_r79345837%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%27m%20not%20sure%20in%20api.one%20is%20safe%20for%20onchange...%5Cr%5CnIn%20the%20old%20api%20we%20would%20have%20issues%20because%20it%20would%20return%20a%20list%20instead%20of%20a%20dict.%5Cr%5CnAnd%20AFAIK%20using%20api.one%20is%20discouraged%20by%20the%20R%26D%20team%22%2C%20%22created_at%22%3A%20%222016-09-19T08%3A35%3A31Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1246629%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dreispt%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20business_requirement_deliverable_crm/models/crm.py%3AL1-32%22%7D%2C%20%22Pull%2032c991a5fd1ff2ca7dd38044bc217312fdcfb2db%20business_requirement_crm/wizard/crm_make_sale.py%2051%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/8%23discussion_r68704201%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22IMO%2C%20by%20experience%2C%20description%20needs%20to%20include%20the%20BR%20name%20as%20well.%5Cr%5Cnsomething%20like%20%5C%22business%20requirement%20name%20/%20deliverable%20name%5C%22%22%2C%20%22created_at%22%3A%20%222016-06-28T06%3A40%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20business_requirement_crm/wizard/crm_make_sale.py%3AL1-65%22%7D%2C%20%22Pull%2032c991a5fd1ff2ca7dd38044bc217312fdcfb2db%20business_requirement_crm/wizard/crm_make_sale.py%2050%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/OCA/business-requirement/pull/8%23discussion_r68704110%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22deprecated.%5Cr%5CnMake%20sure%20to%20get%20the%20taxes%20from%20the%20product%20including%20the%20fiscal%20position%20%28check%20standard%20code%20CrmMakeSales%29%22%2C%20%22created_at%22%3A%20%222016-06-28T06%3A38%3A41Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7600613%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/elicoidal%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20business_requirement_crm/wizard/crm_make_sale.py%3AL1-65%22%7D%7D%7D'></a>
  <a href='https://www.codereviewhub.com/'></a>
  - [ ] <a href='#crh-comment-Pull 32c991a5fd1ff2ca7dd38044bc217312fdcfb2db business_requirement_crm/wizard/crm_make_sale.py 50'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/8#discussion_r68704110'>File: business_requirement_crm/wizard/crm_make_sale.py:L1-65</a></b>
  - <a href='https://github.com/elicoidal'></a> deprecated.
    Make sure to get the taxes from the product including the fiscal position (check standard code CrmMakeSales)
  - [ ] <a href='#crh-comment-Pull 32c991a5fd1ff2ca7dd38044bc217312fdcfb2db business_requirement_crm/wizard/crm_make_sale.py 51'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/8#discussion_r68704201'>File: business_requirement_crm/wizard/crm_make_sale.py:L1-65</a></b>
  - <a href='https://github.com/elicoidal'></a> IMO, by experience, description needs to include the BR name as well.
    something like "business requirement name / deliverable name"
  - [ ] <a href='#crh-comment-Pull 32c991a5fd1ff2ca7dd38044bc217312fdcfb2db business_requirement_crm/models/business.py 22'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/8#discussion_r74006969'>File: business_requirement_crm/models/business.py:L1-32</a></b>
  - <a href='https://github.com/victormartinelicocorp'></a> @elicoidal This is also deprecated the `br_ids` feature is no longer introduced by base module `business_requirement`. I guess I should make this module depends on the `business_requirement_project`. What do you think?
  - <a href='https://github.com/elicoidal'></a> AFAIR, project dependency has been left in business_requirement module no?
    I dont think this is a good idea to this particular module to depend on br_project (which is specific for project management.
    I think this module should depend on br_deliverable so you might want to add some fields in br_deliverable if necessary (or add them in this module)
  - <a href='https://github.com/victormartinelicocorp'></a> Is also a little far for me in memory. But yes I remember that we decided left master_project on the business_requirement.
    Then we might need to modify project to add one2many relation to the `business.requirement` so I'll do as you suggest and modify the br_deliverable.
    After that I'll rebase this PR.
  - [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/8#issuecomment-246562807'>General Comment</a></b>
  - <a href='https://github.com/coveralls'></a> [Coverage Status](https://coveralls.io/builds/7856217)
    Coverage remained the same at 80.0% when pulling **9ca75b8ad6a4edd84e8bffd3623a564a8cd192f0 on victormartinelicocorp:6_business_requirement_crm** into **af25af1b54bfc4fa2f871f713be1a340bc62aa69 on OCA:8.0**.
  - <a href='https://github.com/coveralls'></a> [Coverage Status](https://coveralls.io/builds/7860396)
    Coverage decreased (-11.0%) to 80.0% when pulling **62ee1a923a26fbc1e8e38bb47965d18208b91756 on victormartinelicocorp:6_business_requirement_crm** into **a59878447383fc5ef273ad0b54fb2f95cb7ba3f9 on OCA:8.0**.
  - <a href='https://github.com/dreispt'></a> Needs a cleanup on the README, like in the other modules.
  - <a href='https://github.com/elicoidal'></a> Yes: PR is pending in the fork. We will get to it this week end (almost there!)
  - <a href='https://github.com/victormartinelicocorp'></a> @elicoidal
  - <a href='https://github.com/victormartinelicocorp'></a> @elicoidal @dreispt
  - [ ] <a href='#crh-comment-Pull ae8704aac1972580699c69be6f8345e5237a9807 business_requirement_deliverable_crm/models/crm.py 27'></a> &nbsp;<b><a href='https://github.com/OCA/business-requirement/pull/8#discussion_r79345837'>File: business_requirement_deliverable_crm/models/crm.py:L1-32</a></b>
  - <a href='https://github.com/dreispt'></a> I'm not sure in api.one is safe for onchange...
    In the old api we would have issues because it would return a list instead of a dict.
    And AFAIK using api.one is discouraged by the R&D team

<a href='https://www.codereviewhub.com/OCA/business-requirement/pull/8?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/OCA/business-requirement/pull/8?approve=1'></a>&nbsp;<a href='https://github.com/OCA/business-requirement/pull/8'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: business_requirement_crm/wizard/crm_make_sale.py
Comment (elicoidal) [inline business_requirement_crm/wizard/crm_make_sale.py]: deprecated.
Make sure to get the taxes from the product including the fiscal position (check standard code CrmMakeSales)

File: business_requirement_crm/wizard/crm_make_sale.py
Comment (elicoidal) [inline business_requirement_crm/wizard/crm_make_sale.py]: IMO, by experience, description needs to include the BR name as well.
something like "business requirement name / deliverable name"

File: business_requirement_crm/models/business.py
Comment (martintorresvm) [inline business_requirement_crm/models/business.py]: @elicoidal This is also deprecated the `br_ids` feature is no longer introduced by base module `business_requirement`. I guess I should make this module depends on the `business_requirement_project`. What do you think?

Comment (elicoidal) [inline business_requirement_crm/models/business.py]: AFAIR, project dependency has been left in business_requirement module no?
I dont think this is a good idea to this particular module to depend on br_project (which is specific for project management.
 I think this module should depend on br_deliverable so you might want to add some fields in br_deliverable if necessary (or add them in this module)

Comment (martintorresvm) [inline business_requirement_crm/models/business.py]: Is also a little far for me in memory. But yes I remember that we decided left master_project on the business_requirement. 
Then we might need to modify project to add one2many relation to the `business.requirement` so I'll do as you suggest and modify the br_deliverable.
After that I'll rebase this PR.

**Comment by dreispt** [comment]
Needs a cleanup on the README, like in the other modules.

**Comment by elicoidal** [comment]
Yes: PR is pending in the fork. We will get to it this week end (almost there!)

**Comment by martintorresvm** [comment]
@elicoidal

**Comment by martintorresvm** [comment]
@elicoidal @dreispt

File: business_requirement_deliverable_crm/models/crm.py
Comment (dreispt) [inline business_requirement_deliverable_crm/models/crm.py]: Replace "_compute_get_" with "_compute" ?

File: business_requirement_deliverable_crm/models/crm.py
Comment (dreispt) [inline business_requirement_deliverable_crm/models/crm.py]: It's safe to use `self.project_id.br_ids` directly.

File: business_requirement_deliverable_crm/models/crm.py
Comment (dreispt) [inline business_requirement_deliverable_crm/models/crm.py]: I'm not sure in api.one is safe for onchange...
In the old api we would have issues because it would return a list instead of a dict.
And AFAIK using api.one is discouraged by the R&D team

File: business_requirement_deliverable_crm/wizard/crm_make_sale.py
Comment (dreispt) [inline business_requirement_deliverable_crm/wizard/crm_make_sale.py]: osv? It's deprecated.

File: business_requirement_deliverable_crm/wizard/crm_make_sale.py
Comment (dreispt) [inline business_requirement_deliverable_crm/wizard/crm_make_sale.py]: Use `openerp.exception.ValidationError`

**Review by dreispt — CHANGES_REQUESTED** [review]

File: business_requirement_deliverable_crm/models/crm.py
Comment (martintorresvm) [reply business_requirement_deliverable_crm/models/crm.py]: @dreispt done

File: business_requirement_deliverable_crm/models/crm.py
Comment (martintorresvm) [reply business_requirement_deliverable_crm/models/crm.py]: @dreispt done

File: business_requirement_deliverable_crm/models/crm.py
Comment (martintorresvm) [reply business_requirement_deliverable_crm/models/crm.py]: @dreispt changed by api.multi

File: business_requirement_deliverable_crm/wizard/crm_make_sale.py
Comment (martintorresvm) [reply business_requirement_deliverable_crm/wizard/crm_make_sale.py]: @dreispt done

File: business_requirement_deliverable_crm/wizard/crm_make_sale.py
Comment (martintorresvm) [reply business_requirement_deliverable_crm/wizard/crm_make_sale.py]: @dreispt done

**Comment by OSevangelist** [comment]
I reviewed the code and the functionality and to me it seems to work as expected and the change request from @dreispt have been integrated properly, so +1 from my side

**Comment by OSevangelist** [comment]
Unfortunately there is still the bug that the BR Deliverable lines don't pull the tax position (VAT) if the CRM Quotation Wizard is invoked. This needs to be fixed before merging

**Comment by elicoidal** [comment]
Merging for the code sprint

**Comment by elicoidal** [comment]
@OSevangelist you are right: pending fixing ( @mmalorni ).

**Pull request merged**
